(Docs) Add Bolt example (#1421)
puppetlabs/bolt: documentation/bolt_examples.md

    ![](bolt_PS_3.png)

    **Note:** `--nodes windows` refers to the name of the group of target nodes that you specified in your inventory file. For more information, see [Specify target nodes](bolt_options.md#).

## Deploy a package with Bolt and Chocolatey

> You can use Bolt with Chocolatey to deploy a package on a Windows node by writing a Puppet plan that installs Chocolatey and uses the Chocolatey package provider to install a package. This is all done using content from the Puppet Forge.

Reviewer: ```suggestion
You can use Bolt with Chocolatey to deploy a package on a Windows node by writing a [Bolt plan](writing_plans.md) that installs Chocolatey and uses Puppet's Chocolatey package provider to install a package. This is all done using content from the [Puppet Forge](https://forge.puppet.com).
```

**Before you begin** 

- Install Bolt and Puppet Development Kit (PDK).
- Ensure you have Powershell and Windows Remote Management (WinRM) access.

In this example, you:

- Build a project-specific configuration using a Bolt project directory and PDK.
- Download module content from the Puppet Forge.
- Write a Bolt plan to apply Puppet code and orchestrate the deployment of a package resource using the Chocolatey provider.

> ### 1. Build a project-specific configuration using a Bolt project directory because Bolt runs in the context of a project directory. This directory contains all of the configuration, code, and data loaded by Bolt.

Reviewer: ```suggestion
### 1. Build a project-specific configuration
Bolt runs in the context of a [Bolt project directory](bolt_project_directories.md). This directory contains all of the configuration, code, and data loaded by Bolt.
```

> 1. Create a directory called `puppet_choco_tap` and add a bolt.yaml file. Whenever you run Bolt inside this project directory, it will load the code and configuration.

Reviewer: ```suggestion
1. Create a directory called `puppet_choco_tap` and add a `bolt.yaml` file. Whenever you run Bolt inside this project directory, it will load the code and configuration.
```

Reviewer: It'd be nice if, below this line, there were the specific commands to run to create the directory.

```
mkdir puppet_choco_tap
New-Item -Type File -Path .\puppet_choco_tap\bolt.yaml
```

> 2. Create an inventory file to store information about your nodes. This is stored as `inventory.yaml` by default in the project directory. Add the following code: 

Reviewer: ```suggestion
2. Create an inventory file to store information about your targets. This is stored as `inventory.yaml` by default in the project directory. Add the following code: 
```

```
groups:
  - name: windows

>     nodes:

Reviewer: ```suggestion
    targets:
```

      - chocowin0.classroom.puppet.com
      - chocowin1.classroom.puppet.com
    config:
      transport: winrm
      winrm:
        user: Administrator
        password: <ADD PASSWORD>
```

> 4. To make sure that your inventory is configured correctly and that you can connect to all the target nodes, run the following command from inside the project directory: 

Reviewer: ```suggestion
4. To make sure that your inventory is configured correctly and that you can connect to all the targets, run the following command from inside the project directory: 
```

```
bolt command run 'echo hi' --targets windows
```

> **Note:** The `--targets windows` argument refers to the node group defined in the inventory file.

Reviewer: ```suggestion
**Note:** The `--targets windows` argument refers to the target group defined in the inventory file.
```

You should get the following output:

```
Started on x.x.x.x...
Started on localhost...
Finished on localhost:
  STDOUT:
    hi
Finished on 0.0.0.0:
  STDOUT:
    hi
Finished on 127.0.0.1:
  STDOUT:
    hi

> Successful on 3 nodes: 0.0.0.0:20022,localhost

Reviewer: ```suggestion
Successful on 3 targets: 0.0.0.0:20022,localhost
```

> Ran on 3 nodes in 0.20 seconds

Reviewer: ```suggestion
Ran on 3 targets in 0.20 seconds
```

```

> ### 2. Download the Chocolatey module and its dependencies from the Puppet Forge. 

Reviewer: ```suggestion
### 2. Download the Chocolatey module
```

> Bolt uses a `Puppetfile` to install module content from the Forge. A `Puppetfile` is a formatted text file that specifies the modules and data you want in each environment.

Reviewer: ```suggestion
Bolt uses a [Puppetfile](https://puppet.com/docs/pe/latest/puppetfile.html) to install module content from the Forge. A `Puppetfile` is a formatted text file that specifies the modules and data you want in each environment.
```

1 . Create a file named `Puppetfile` in the project directory, with the modules needed for this example:

```     
# Modules from the Puppet Forge
mod 'puppetlabs-chocolatey', '4.1.0'
## dependencies
mod 'puppetlabs-stdlib', '4.13.1' #install latest
mod 'puppetlabs-powershell', '2.3.0'
mod 'puppetlabs-registry', '2.1.0'
        
# Modules from Git
# mod 'apache',
# :git    => 'https://github.com/puppetlabs/puppetlabs-apache',
# :commit => 'de290646f97e04b4b8e42c70f6e01e860c394ce7'     
 ```

> Note that you can install modules from a number of different sources. For more information, see the [Puppetfile readme](https://github.com/puppetlabs/r10k/blob/master/doc/puppetfile.mkd#examples).

Reviewer: ```suggestion
Note that you can install modules from a number of different sources. For more information, see the [Puppetfile README](https://github.com/puppetlabs/r10k/blob/master/doc/puppetfile.mkd#examples).
```

2. Add the Puppetfile to a remote repository, for example, GitHub.

3. From inside the project directory, install the required modules:

```
bolt puppetfile install
```

After it runs, you can see a `modules` directory inside the project directory, containing the modules you specified in the `Puppetfile`.

> ### 3. Write a Bolt plan to apply Puppet code and orchestrate the deployment of a package resource using the Chocolatey provider. Plans allow you to run more than one task with a single command.

Reviewer: ```suggestion
### 3. Write a Bolt plan to apply Puppet code

Write a Bolt plan to orchestrate the deployment of a package resource using the Chocolatey provider. Plans allow you to run more than one task with a single command, compute values for the input to a task, process the results of tasks, or make decisions based on the result of running a task.
```

1. Create a `site-modules` directory. This is where you will add code and local modules.

2. Inside the `site-modules` directory, create a new module called `puppet_choco_tap`.

3. Inside the `puppet_choco_tap` module, create a plans directory with an `init.pp file` called `/plans/installer.pp`.

The folder tree should look like this:

```

> ├──plans

Reviewer: I would do this one level higher, so it's:
```
site-modules
└── puppet_choco_tap
    └── plans
        └── installer.pp
```

   └── installer.pp 
```

> 2. Create a plan called `puppet_chocolatey_tap::installer` by copying the following code into the `installer.pp` file::

Reviewer: ```suggestion
2. Create a plan called `puppet_choco_tap::installer` by copying the following code into the `installer.pp` file::
```

```
plan puppet_choco_tap::installer(

>   TargetSpec $nodes,

Reviewer: ```suggestion
  TargetSpec $targets,
```

  String $package,
  Variant[Enum['absent', 'present'], String ] $ensure = 'present',
){

>   apply_prep($nodes)

Reviewer: ```suggestion
  apply_prep($targets)
```

>   apply($nodes){

Reviewer: ```suggestion
  apply($targets){
```

    include chocolatey

    package { $package :
      ensure    => $ensure,
      provider  => 'chocolatey',
      }
    }
  }
}
```
Take note of the following features of the plan:

> - It has three parameters: the `TargetSpec`, a `package` string for the package name, and the `ensure` state of the package which allows for version, absent or present.

Reviewer: ```suggestion
- It has three parameters: the list of targets to install the package on, a `package` string for the package name, and the `ensure` state of the package which allows for version, absent or present.
```

> - It has the `apply_prep` function call, which is used to install modules needed by `apply` on remote nodes as well as to gather facts about the nodes.

Reviewer: ```suggestion
- It has the `apply_prep` function call, which is used to install modules needed by `apply` on targets as well as to gather facts about the targets.
```

> - The first `apply` block installs the Chocolatey package manager. The Chocolatey provider is also deployed as a library with the Puppet agent in `apply_prep`.

Reviewer: ```suggestion
- `include chocolatey` installs the Chocolatey package manager. The Chocolatey provider is also deployed as a library with the Puppet agent in `apply_prep`.
```

> - The second `apply` block installs a specified package using the Chocolatey provider.

Reviewer: ```suggestion
- The [package resource](https://puppet.com/docs/puppet/latest/types/package.html) ensures a package's state using the Chocolatey provider.
```

2. To verify that the `puppet_choco_tap::installer` plan is available, add and commit your code to your remote repository, and run the following command again:

`bolt puppetfile install`

> Then run:

Reviewer: ```suggestion
```

Finished: install puppet and gather facts with 0 failures in 22.11 sec
Starting: apply catalog on chocowin0.classroom.puppet.com, chocowin1.classroom.puppet.com
Finished: apply catalog with 0 failures in 18.77 sec
Starting: apply catalog on chocowin0.classroom.puppet.com, chocowin1.classroom.puppet.com
Finished: apply catalog with 0 failures in 33.74 sec
Finished: plan puppet_choco_tap::installer in 74.63 sec
```

4. To check that the installation worked, run the following `frogsay` command: 

```        
bolt command run 'frogsay ribbit' --targets=windows 
```

The result will vary on each server, and will look something like this:

```
Started on chocowin1.classroom.puppet.com...
Started on chocowin0.classroom.puppet.com...
Finished on chocowin0.classroom.puppet.com:
  STDOUT:
    
            DO NOT PAINT OVER FROG.
            /
      @..@
     (----)
    ( >__< )
    ^^ ~~ ^^
Finished on chocowin1.classroom.puppet.com:
  STDOUT:
    
            TO PREVENT THE RISK OF FIRE OR ELECTRIC SHOCK, DO NOT ENGAGE WITH FROG
            WHILE AUTOMATIC UPDATES ARE BEING INSTALLED.
            /
      @..@
     (----)
    ( >__< )
    ^^ ~~ ^^

> Successful on 2 nodes: chocowin0.classroom.puppet.com,chocowin1.classroom.puppet.com

Reviewer: ```suggestion
Successful on 2 targets: chocowin0.classroom.puppet.com,chocowin1.classroom.puppet.com
```

> Ran on 2 nodes in 3.15 seconds

Reviewer: ```suggestion
Ran on 2 targets in 3.15 seconds
```

```

> That’s it! In this one plan, you have both installed Chocolatey and deployed the package to two nodes. You can do the same thing on any number of nodes by editing the inventory file. Note that Chocolatey will remain installed on your machine.

Reviewer: ```suggestion
That’s it! In this one plan, you have both installed Chocolatey and deployed the package to two targets. You can do the same thing on any number of nodes by editing the inventory file. Note that Chocolatey will remain installed on your machine.
```

    ![](bolt_PS_3.png)

    **Note:** `--nodes windows` refers to the name of the group of target nodes that you specified in your inventory file. For more information, see [Specify target nodes](bolt_options.md#).

## Deploy a package with Bolt and Chocolatey

You can use Bolt with Chocolatey to deploy a package on a Windows node by writing a Puppet plan that installs Chocolatey and uses the Chocolatey package provider to install a package. This is all done using content from the Puppet Forge.

**Before you begin** 

- Install Bolt and Puppet Development Kit (PDK).
- Ensure you have Powershell and Windows Remote Management (WinRM) access.

In this example, you:

- Build a project-specific configuration using a Bolt project directory and PDK.
- Download module content from the Puppet Forge.
- Write a Bolt plan to apply Puppet code and orchestrate the deployment of a package resource using the Chocolatey provider.

> ### 1. Build a project-specific configuration using a Bolt project directory Bolt runs in the context of a project directory. This directory contains all of the configuration, code, and data loaded by Bolt.

Reviewer 2: ```suggestion
### 1. Build a project-specific configuration using a Bolt project directory Bolt runs in the context of a project directory. This directory contains all of the configuration, code, and data loaded by Bolt.
```

> ### 1. Build a project-specific configuration using a Bolt project directory Bolt runs in the context of a project directory. This directory contains all of the configuration, code, and data loaded by Bolt.

Reviewer 2: It seems as if there's missing punctuation or missing conjunction in this header.

> ### 1. Build a project-specific configuration using a Bolt project directory Bolt runs in the context of a project directory. This directory contains all of the configuration, code, and data loaded by Bolt.

Reviewer 2: ```suggestion
### 1. Build a project-specific configuration using a Bolt project directory because Bolt runs in the context of a project directory. This directory contains all of the configuration, code, and data loaded by Bolt.
```

The folder tree should look like this:

```
├──plans
   └── installer.pp 
```

2. Create a plan called `puppet_chocolatey_tap::installer` by copying the following code into the `installer.pp` file::

```
plan puppet_choco_tap::installer(
  TargetSpec $nodes,
  String $package,
  Variant[Enum['absent', 'present'], String ] $ensure = 'present',
){
  apply_prep($nodes)

  apply($nodes){
    include chocolatey

    package { $package :
      ensure    => $ensure,
      provider  => 'chocolatey',
      }
    }
  }
}
```
Take note of the following features of the plan:

- It has three parameters: the `TargetSpec`, a `package` string for the package name, and the `ensure` state of the package which allows for version, absent or present.
- It has the `apply_prep` function call, which is used to install modules needed by `apply` on remote nodes as well as to gather facts about the nodes.
- The first `apply` block installs the Chocolatey package manager. The Chocolatey provider is also deployed as a library with the Puppet agent in `apply_prep`.
- The second `apply` block installs a specified package using the Chocolatey provider.

> 2. To verify that the `puppet_choco_tap::installer` plan is available, add and commit your code to your remote repository, and run the following command again:

Reviewer 3: You will not need to commit code and re-run puppetfile install. 
```suggestion
2. To verify that the `puppet_choco_tap::installer` plan is available:
```
then remove `bolt puppetfile install` and keep `bolt plan show`.

Reviewer: @clairecadman Same here - Please add Cas' update

    **Note:** `--nodes windows` refers to the name of the group of target nodes that you specified in your inventory file. For more information, see [Specify target nodes](bolt_options.md#).

## Deploy a package with Bolt and Chocolatey

You can use Bolt with Chocolatey to deploy a package on a Windows node by writing a Puppet plan that installs Chocolatey and uses the Chocolatey package provider to install a package. This is all done using content from the Puppet Forge.

**Before you begin** 

- Install Bolt and Puppet Development Kit (PDK).
- Ensure you have Powershell and Windows Remote Management (WinRM) access.

In this example, you:

- Build a project-specific configuration using a Bolt project directory and PDK.
- Download module content from the Puppet Forge.
- Write a Bolt plan to apply Puppet code and orchestrate the deployment of a package resource using the Chocolatey provider.

### 1. Build a project-specific configuration using a Bolt project directory Bolt runs in the context of a project directory. This directory contains all of the configuration, code, and data loaded by Bolt.
    
1. Create a module using the PDK command `pdk new module puppet_choco_tap` 

> 2. To make the module a Bolt project directory, add a `bolt.yaml` file. 

Reviewer 4: I don't think we should promote this workflow. It works for some cases but breaks the "things behave properly as long as your inside your project" since a module will exist inside another project eventually.

What does creating a module with PDK buy us?

```
mkdir puppet_choco_tap
```

> 2. Add a `bolt.yaml` file to the `puppet_choco_tap` directory:

Author: @lucywyman could you please verify. Thanks!

        password: <ADD PASSWORD>
```

4. To make sure that your inventory is configured correctly and that you can connect to all the targets, run the following command from inside the project directory: 

```
bolt command run 'echo hi' --targets windows
```

**Note:** The `--targets windows` argument refers to the target group defined in the inventory file.

You should get the following output:

```
Started on x.x.x.x...
Started on localhost...
Finished on localhost:
  STDOUT:
    hi
Finished on 0.0.0.0:
  STDOUT:
    hi
Finished on 127.0.0.1:
  STDOUT:
    hi
Successful on 3 targets: 0.0.0.0:20022,localhost
Ran on 3 targets in 0.20 seconds
```

### 2. Download the Chocolatey module

Bolt uses a [Puppetfile](https://puppet.com/docs/pe/latest/puppetfile.html) to install module content from the Forge. A `Puppetfile` is a formatted text file that specifies the modules and data you want in each environment.

1 . Create a file named `Puppetfile` in the project directory, with the modules needed for this example:

```     
mod 'puppetlabs-chocolatey', '4.1.0'
mod 'puppetlabs-stdlib', '4.13.1'
mod 'puppetlabs-powershell', '2.3.0'
mod 'puppetlabs-registry', '2.1.0'

> mod 'apache'

Author: @lucywyman like this?

    hi
Successful on 3 nodes: 0.0.0.0:20022,localhost
Ran on 3 nodes in 0.20 seconds
```

### 2. Download the Chocolatey module and its dependencies from the Puppet Forge. 

Bolt uses a `Puppetfile` to install module content from the Forge. A `Puppetfile` is a formatted text file that specifies the modules and data you want in each environment.

1 . Create a file named `Puppetfile` in the project directory, with the modules needed for this example:

```     
# Modules from the Puppet Forge
mod 'puppetlabs-chocolatey', '4.1.0'
## dependencies
mod 'puppetlabs-stdlib', '4.13.1' #install latest
mod 'puppetlabs-powershell', '2.3.0'
mod 'puppetlabs-registry', '2.1.0'
        
# Modules from Git
# mod 'apache',
# :git    => 'https://github.com/puppetlabs/puppetlabs-apache',
# :commit => 'de290646f97e04b4b8e42c70f6e01e860c394ce7'     
 ```

Note that you can install modules from a number of different sources. For more information, see the [Puppetfile readme](https://github.com/puppetlabs/r10k/blob/master/doc/puppetfile.mkd#examples).

2. Add the Puppetfile to a remote repository, for example, GitHub.

3. Add the following code to the Puppetfile — with your git user account: 

```     

> mod 'puppet_choco_tap',

Reviewer 3: I don't think you want this anymore.

    :git    =>  'https://github.com/<your git user>/puppet_choco_tap.git'
```

4. From inside the project directory, install the required modules:

```
bolt puppetfile install
```

After it runs, you can see a `modules` directory inside the project directory, containing the modules you specified in the `Puppetfile`.

### 3. Write a Bolt plan to apply Puppet code and orchestrate the deployment of a package resource using the Chocolatey provider. Plans allow you to run more than one task with a single command.

> 1. Create a `site-modules` directory. This is where you will add code and local modules.

Author: @adreyer What do you mean by local modules?

Reviewer 4: modules that are committed directly to the project repository instead of managed by the Pupeptfile as seperate git repos or forge modules

```
plan puppet_choco_tap::installer(
  TargetSpec $nodes,
  String $package,
  Variant[Enum['absent', 'present'], String ] $ensure = 'present',
){
  apply_prep($nodes)

  apply($nodes){
    include chocolatey

    package { $package :
      ensure    => $ensure,
      provider  => 'chocolatey',
      }
    }
  }
}
```
Take note of the following features of the plan:

- It has three parameters: the `TargetSpec`, a `package` string for the package name, and the `ensure` state of the package which allows for version, absent or present.
- It has the `apply_prep` function call, which is used to install modules needed by `apply` on remote nodes as well as to gather facts about the nodes.
- The first `apply` block installs the Chocolatey package manager. The Chocolatey provider is also deployed as a library with the Puppet agent in `apply_prep`.
- The second `apply` block installs a specified package using the Chocolatey provider.

2. To verify that the `puppet_choco_tap::installer` plan is available, add and commit your code to your remote repository, and run the following command again:

> `bolt puppetfile deploy`

Reviewer 3: ```suggestion
`bolt puppetfile install`
```

### 1. Build a project-specific configuration using a Bolt project directory Bolt runs in the context of a project directory. This directory contains all of the configuration, code, and data loaded by Bolt.
    
1. Create a module using the PDK command `pdk new module puppet_choco_tap` 

2. To make the module a Bolt project directory, add a `bolt.yaml` file. 

3. Add the following configuration to the `bolt.yaml` file. 

```
Modulepath: "./modules/:~/.puppetlabs/bolt-code/modules:~/.puppetlabs/bolt-code/site-modules"
```

This means that the project is going to support itself by deploying all module dependencies to a modules folder in the project. Note that you don’t have to create the `./modules` folder.

4. Create an inventory file to store information about your nodes. This is stored as `inventory.yaml` by default in the project directory. Add the following code: 

```
groups:
  - name: windows
    nodes:
      - chocowin0.classroom.puppet.com
      - chocowin1.classroom.puppet.com
    config:
      transport: winrm
      winrm:
        user: Administrator

>         Password: <ADD PASSWORD>

Reviewer 3: ```suggestion
        password: <ADD PASSWORD>
```

```

5. To make sure that your inventory is configured correctly and that you can connect to all the target nodes, run the following command from inside the project directory: 

```
bolt command run 'echo hi' --targets windows
```

**Note:** The `--targets windows` argument refers to the node group defined in the inventory file.

You should get the following output:

```
Started on x.x.x.x...
Started on localhost...
Finished on localhost:
  STDOUT:
    hi
Finished on 0.0.0.0:
  STDOUT:
    hi
Finished on 127.0.0.1:
  STDOUT:
    hi
Successful on 3 nodes: 0.0.0.0:20022,localhost
Ran on 3 nodes in 0.20 seconds
```

### 2. Download the Chocolatey module and its dependencies from the Puppet Forge. 

Bolt uses a `Puppetfile` to install module content from the Forge. A `Puppetfile` is a formatted text file that specifies the modules and data you want in each environment.

1. Create a file named `Puppetfile` in the project directory — which describes the Forge content you want to install: 

```     
# Modules from the Puppet Forge
mod 'puppetlabs-chocolatey', '4.1.0'
## dependencies
mod 'puppetlabs-stdlib', '4.13.1' #install latest

>  mod 'puppetlabs-powershell', '2.3.0'

Reviewer 3: Do not indent here.

mod 'puppetlabs-registry', '2.1.0'
        

> # Modules from Git

Reviewer 3: I think this commented out code is more confusing than it is worth. can we link (maybe to r10k readme) on how to do this?

Reviewer 3: I still think we should delete the commented out code. We can link to the reference docs instead.

Reviewer: @clairecadman I agree, and I think this still needs to be addressed

# Examples: https://github.com/puppetlabs/r10k/blob/master/doc/puppetfile.mkd#examples
#mod 'apache',
#  :git    => 'https://github.com/puppetlabs/puppetlabs-apache',
#  :commit => 'de290646f97e04b4b8e42c70f6e01e860c394ce7'     
 ```

2. Add the Puppetfile to a version control service, for example, Github.

3. Add the following code to the Puppetfile — with your git user account: 

```     
mod 'puppet_choco_tap',
    :git    =>  'https://github.com/<your git user>/puppet_choco_tap.git'
```

4. From inside the project directory, install the Forge content by running: 

```
bolt puppetfile install
```

After it runs, you can see a `modules` directory inside the project directory, containing the modules you specified in the `Puppetfile`.

### 3. Write a Bolt plan to apply Puppet code and orchestrate the deployment of a package resource using the Chocolatey provider. Plans allow you to run more than one task with a single command.

> 1. Inside your project directory, in a file called `/plan/installer.pp`, create a plan called `puppet_chocolatey_tap::installer`. 

Reviewer 3: Plan must be in a `plans` directory, not `plan`.

Reviewer 4: also it must be inside a module not a project.

The folder tree should look like this:

```

> ├──plan

Reviewer 3: plans

   └── installer.pp 
```

2. Copy the following code into the `installer.pp` file:

```
plan puppet_choco_tap::installer(
   TargetSpec                                   $nodes,
   String                                       $package,
   Variant[Enum['absent', 'present'], String ]  $ensure = 'present',
 ){

 apply_prep($nodes)

apply($nodes){
  include chocolatey
 }
apply($nodes){
  package { $package :
    ensure    => $ensure,
    provider  => 'chocolatey',
    }
  }
}
```

Take note of the following features of the plan:

- It has three parameters: the `TargetSpec`, a `package` string for the package name, and the `ensure` state of the package which allows for version, absent or present.
- It has the `apply_prep` function call, which is used to install modules needed by `apply` on remote nodes as well as to gather facts about the nodes.
- The first `apply` block installs the Chocolatey package manager. The Chocolatey provider is also deployed as a library with the Puppet agent in `apply_prep`.
- The second `apply` block installs a specified package using the Chocolatey provider.

> 2. To verify that the `puppet_chocolatey_tap` plan is available, run the following command: 

Reviewer 3: With the workflow described here, you would have to first check in the code to your remote repository, then re-run the puppetfile install. I would recommend though, describing checking in the code here and adding the line to the puppetfile here. As written the plan will not be in the correct location because r10k (bolt puppetfile install) has not installed the module under `modules`.

Author: @donoghuc I've made all the other changes you've suggested, thank you!
I'm not too clear on what you mean here though. Could you please at this step?

Reviewer 3: So to me it looks like the way module content ends up in the expected location is for r10k to fetch the code from the remote repo and add it under 'modules/puppetlabs/plans'. So any change you want to make you must push to remote and then re-install with `bolt puppetfile install`. I think that we should explicitly describe this model if that is how the author wants people to use this workflow. In addiition to that I think we need to have the instructions explicitly add the steps for checking in the newly minted plan into the remote repo and then pulling the change with r10k (bolt puppetfile install). Otherwise this will not work as expected bc the plan will not be found.

Author: @donoghuc Let me know how this looks, thanks!

```
bolt plan show
```

The output looks like this:

```
facts
facts::info
puppet_choco_tap::installer
puppetdb_fact
```

3. Run the plan with the `bolt plan run` command: 

```

> bolt plan run puppet_choco_tap::installer package=frogsay --nodes=win --password

Reviewer 3: Wrong target for inventory above. Should be windows, also do not prompt for password bc it is in the inventoryfile.

Reviewer 3: Also, nit, but use `--targets` not `--nodes`.

```

The output looks like this:

```
Starting: plan puppet_choco_tap::installer
Starting: install puppet and gather facts on chocowin0.classroom.puppet.com, chocowin1.classroom.puppet.com
Finished: install puppet and gather facts with 0 failures in 22.11 sec
Starting: apply catalog on chocowin0.classroom.puppet.com, chocowin1.classroom.puppet.com
Finished: apply catalog with 0 failures in 18.77 sec
Starting: apply catalog on chocowin0.classroom.puppet.com, chocowin1.classroom.puppet.com
Finished: apply catalog with 0 failures in 33.74 sec
Finished: plan puppet_choco_tap::installer in 74.63 sec
```

4. To check that the installation worked, run the following `frogsay` command: 

```        

> bolt command run 'frogsay ribbit' --targets windows --password

Reviewer 3: Do not prompt for password. Given it is in the inventoryfile it would be ignored anyway.

## Deploy a package with Bolt and Chocolatey

You can use Bolt with Chocolatey to deploy a package on a Windows node by writing a Puppet plan that installs Chocolatey and uses the Chocolatey package provider to install a package. This is all done using content from the Puppet Forge.

**Before you begin** 

- Install Bolt and Puppet Development Kit (PDK).
- Ensure you have Powershell and Windows Remote Management (WinRM) access.

In this example, you:

- Build a project-specific configuration using a Bolt project directory and PDK.
- Download module content from the Puppet Forge.
- Write a Bolt plan to apply Puppet code and orchestrate the deployment of a package resource using the Chocolatey provider.

### 1. Build a project-specific configuration using a Bolt project directory Bolt runs in the context of a project directory. This directory contains all of the configuration, code, and data loaded by Bolt.
    
1. Create a module using the PDK command `pdk new module puppet_choco_tap` 

2. To make the module a Bolt project directory, add a `bolt.yaml` file. 

3. Add the following configuration to the `bolt.yaml` file. 

```

> Modulepath: "./modules/:~/.puppetlabs/bolt-code/modules:~/.puppetlabs/bolt-code/site-modules"

Reviewer 3: This seems wrong. Why is `Modulepath` capitialized? I'm thinking maybe this appears to work becuase `Modulepath` gets ignored. It looks like everything should be contained within the `puppet_choco_tap` module (bolt project dir) in which case all you would need is an empty `bolt.yaml`. So setting the modulepath would be unnecessary and i *think* would not find the project specific plan.

Author: @donoghuc Do you think this whole step should be removed?

Reviewer 3: Yeah, it is not needed. I just ran through to verify. As it is written Modulepath is ignored.

Reviewer 5: that's just where the conversion from google doc as picked up the capitalisation it should be lowercase, it is needed for the example workflow.

      - chocowin0.classroom.puppet.com
      - chocowin1.classroom.puppet.com
    config:
      transport: winrm
      winrm:
        user: Administrator
        Password: <ADD PASSWORD>
```

5. To make sure that your inventory is configured correctly and that you can connect to all the target nodes, run the following command from inside the project directory: 

```
bolt command run 'echo hi' --targets windows
```

**Note:** The `--targets windows` argument refers to the node group defined in the inventory file.

You should get the following output:

```
Started on x.x.x.x...
Started on localhost...
Finished on localhost:
  STDOUT:
    hi
Finished on 0.0.0.0:
  STDOUT:
    hi
Finished on 127.0.0.1:
  STDOUT:
    hi
Successful on 3 nodes: 0.0.0.0:20022,localhost
Ran on 3 nodes in 0.20 seconds
```

### 2. Download the Chocolatey module and its dependencies from the Puppet Forge. 

Bolt uses a `Puppetfile` to install module content from the Forge. A `Puppetfile` is a formatted text file that specifies the modules and data you want in each environment.

1. Create a file named `Puppetfile` in the project directory — which describes the Forge content you want to install: 

```

> forge 'https://forge.puppet.com'

Reviewer 4: This line should be removed. The rest of the lines should not be indented.

3. Add the following code to the Puppetfile — with your git user account: 

```     
mod 'puppet_choco_tap',
    :git    =>  'https://github.com/<your git user>/puppet_choco_tap.git'
```

4. From inside the project directory, install the Forge content by running: 

```
bolt puppetfile install
```

After it runs, you can see a `modules` directory inside the project directory, containing the modules you specified in the `Puppetfile`.

### 3. Write a Bolt plan to apply Puppet code and orchestrate the deployment of a package resource using the Chocolatey provider. Plans allow you to run more than one task with a single command.

1. Inside your project directory, in a file called `/plan/installer.pp`, create a plan called `puppet_chocolatey_tap::installer`. 

The folder tree should look like this:

```
├──plan
   └── installer.pp 
```

2. Copy the following code into the `installer.pp` file:

```
plan puppet_choco_tap::installer(
   TargetSpec                                   $nodes,
   String                                       $package,
   Variant[Enum['absent', 'present'], String ]  $ensure = 'present',
 ){

 apply_prep($nodes)

apply($nodes){
  include chocolatey
 }

> apply($nodes){

Reviewer 3: The indentations for the second apply block are funky. Also, I think that given we do not do anything with the result of the first apply statement and apply the manifest code on the same sets of targets, I think we should have a single statement with all the manifest code (bolt performance).

Author: @donoghuc I'm not quite sure what you mean, would you mind updating the code with this?

Reviewer 3: ```puppet
plan puppet_choco_tap::installer(
  TargetSpec $nodes,
  String $package,
  Variant[Enum['absent', 'present'], String ] $ensure = 'present',
){
  apply_prep($nodes)

  apply($nodes){
    include chocolatey

    package { $package :
      ensure    => $ensure,
      provider  => 'chocolatey',
      }
    }
  }
}
```

```
bolt plan run puppet_choco_tap::installer package=frogsay --nodes=win --password
```

The output looks like this:

```
Starting: plan puppet_choco_tap::installer
Starting: install puppet and gather facts on chocowin0.classroom.puppet.com, chocowin1.classroom.puppet.com
Finished: install puppet and gather facts with 0 failures in 22.11 sec
Starting: apply catalog on chocowin0.classroom.puppet.com, chocowin1.classroom.puppet.com
Finished: apply catalog with 0 failures in 18.77 sec
Starting: apply catalog on chocowin0.classroom.puppet.com, chocowin1.classroom.puppet.com
Finished: apply catalog with 0 failures in 33.74 sec
Finished: plan puppet_choco_tap::installer in 74.63 sec
```

4. To check that the installation worked, run the following `frogsay` command: 

```        
bolt command run 'frogsay ribbit' --targets windows --password
```

The result will vary on each server, and will look something like this:

```
Started on chocowin1.classroom.puppet.com...
Started on chocowin0.classroom.puppet.com...
Finished on chocowin0.classroom.puppet.com:
  STDOUT:
    
            DO NOT PAINT OVER FROG.
            /
      @..@
     (----)
    ( >__< )

>     ^^ ~~ ^^

Reviewer 3: :frog:
